### PR TITLE
[SPARK-24884][SQL] add regexp_extract_all support

### PR DESCRIPTION
**Pull request by xueyumusic**
### What changes were proposed in this pull request?

This PR tried to add `regexp_extract_all` sql function. It is supported in [presto](https://prestodb.io/docs/current/functions/regexp.html) and [pig](https://pig.apache.org/docs/latest/api/org/apache/pig/builtin/REGEX_EXTRACT_ALL.html)
Also it refactored regexp_extract since there were some duplicated codes.

### Why are the changes needed?

Added `regexp_extract_all` function

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

Unit test

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 1]
```diff
@@ -446,3 +448,88 @@ case class RegExpExtract(subject: Expression, regexp: Expression, idx: Expressio
...
+ * NOTE: this expression is not THREAD-SAFE, as it has some internal mutable status.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(str, regexp[, idx]) - Extracts all groups that matches `regexp`.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_('100-200,300-400', '(\\d+)-(\\d+)', 1);
+       [100, 300]
+  """)
+case class RegExpExtractAll(subject: Expression, regexp: Expression, idx: Expression)
```
Comment (xuanyuanking) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: Add an abstract class to reduce duplicated code between `RegExpExtractAll` and `RegExpExtract`?

**Comment by kiszk** [comment]
@xueyumusic gentle ping

**Comment by xueyumusic** [comment]
Thanks for ping, @kiszk , I will update the code these days.

**Comment by afsanjar** [comment]
thanks for the patch @xueyumusic, we have been waiting for such feature in Spark for awhile :) any update when your hard work will be publicly available.

**Comment by xueyumusic** [comment]
Hi, @kiszk @afsanjar very sorry for long delay update. I updated to address feedback of @xuanyuanking , thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 2]
```diff
@@ -386,6 +388,41 @@ case class RegExpReplace(subject: Expression, regexp: Expression, rep: Expressio
...
+    val termPattern = ctx.addMutableState(classNamePattern, "pattern")
+
+    val setEvNotNull = if (nullable) {
+      s"${ev.isNull} = false;"
+    } else {
+      ""
+    }
+
+    (classNamePattern, matcher, matchResult, termLastRegex, termPattern, setEvNotNull)
+
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: nit: delete empty line

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 3]
```diff
@@ -471,3 +485,72 @@ case class RegExpExtract(subject: Expression, regexp: Expression, idx: Expressio
...
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val groupArray = ctx.freshName("groupArray")
+    val arrayClass = classOf[GenericArrayData].getName
+    val (classNamePattern, matcher, matchResult, termLastRegex, termPattern, setEvNotNull) =
+      getDoGenCodeVals(ctx, ev)
+
+    nullSafeCodeGen(ctx, ev, (subject, regexp, idx) => {
+      s"""
+      if (!$regexp.equals($termLastRegex)) {
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: nit: could you please use the following pattern?
```
s"""
   |
   |
 """
```

Comment (xueyumusic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: Thanks, @kiszk , updated.

**Comment by maropu** [comment]
We recently updated the PR template: https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE, so could you please follow that, first?

**Comment by maropu** [comment]
ok to test

**Comment by maropu** [comment]
Also, could you check if the other databases support the similar function, e.g., oracle, potgresql, mysql, ..., and then describe the result in the PR description?

**Comment by xueyumusic** [comment]
Hi, @maropu , I updated PR description based on new template, thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 4]
```diff
@@ -386,6 +388,40 @@ case class RegExpReplace(subject: Expression, regexp: Expression, rep: Expressio
...
+  // last regex in string, we will update the pattern iff regexp value changed.
+  @transient protected var lastRegex: UTF8String = _
+  // last regex pattern, we cache it for performance concern
+  @transient protected var pattern: Pattern = _
+  override def inputTypes: Seq[AbstractDataType] = Seq(StringType, StringType, IntegerType)
+
+  protected def getMatcher(s: Any, p: Any) = {
+    if (!p.equals(lastRegex)) {
+      // regex value changed
+      lastRegex = p.asInstanceOf[UTF8String].clone()
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: nit: Do we need to clone `p`?

**Comment by maropu** [comment]
> Also, could you check if the other databases support the similar function, e.g., oracle, potgresql, mysql, ..., and then describe the result in the PR description?

Have you checked this? pig and presto only support this?

**Comment by HyukjinKwon** [comment]
ping @xueyumusic

**Comment by justgfather** [comment]
Hey, was this feature merged in the master. If yes, in what version of spark?

**Comment by kiszk** [comment]
@justgfather Not merged yet. If this would be merged now, you can use this function in Spark 3.0.

**Comment by gatorsmile** [comment]
cc @beliefer Could you take this over?

**Comment by beliefer** [comment]
@gatorsmile OK.

**Pull request closed without merging**